File: SwrveSDK/SDK/Messaging/SwrveEmbeddedCampaign.m
#import "SwrveEmbeddedCampaign.h"
#import "SwrveEmbeddedMessage.h"
#import "SwrveCampaign+Private.h"
#import "SwrveMessageController+Private.h"

@implementation SwrveEmbeddedCampaign

@synthesize message;

-(id)initAtTime:(NSDate *)time fromDictionary:(NSDictionary *)json forController:(SwrveMessageController *)controller {
    id instance = [super initAtTime:time fromDictionary:json];
    NSDictionary *embedded_message = [json objectForKey:@"embedded_message"];
    SwrveEmbeddedMessage *embeddedMessage = [[SwrveEmbeddedMessage alloc] initWithDictionary:embedded_message forCampaign:self forController:controller];
    self.message = embeddedMessage;
    
    if ([json objectForKey:@"subject"] != [NSNull null] && [json objectForKey:@"subject"] != nil) {
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wdeprecated-declarations"
        self.subject = [json objectForKey:@"subject"];
#pragma clang diagnostic pop
    }
    
    self.campaignType = SWRVE_CAMPAIGN_EMBEDDED;
    
    self.name = self.message.name;
    
    return instance;
}

-(BOOL)hasMessageForEvent:(NSString *)event {

    return [self hasMessageForEvent:event withPayload:nil];
}

-(BOOL)hasMessageForEvent:(NSString *)event withPayload:(NSDictionary *)payload {

    return [self canTriggerWithEvent:event andPayload:payload];
}

- (SwrveEmbeddedMessage*)messageForEvent:(NSString *)event
                       withPayload:(NSDictionary *)payload
                            atTime:(NSDate *)time
                       withReasons:(NSMutableDictionary *)campaignReasons {

    if (![self hasMessageForEvent:event withPayload:payload]){

        [SwrveLogger debug:@"There is no trigger in %ld that matches %@", (long)self.ID, event];
        [self logAndAddReason:[NSString stringWithFormat:@"There is no trigger in %ld that matches %@ with conditions %@", (long)self.ID, event, payload] withReasons:campaignReasons];
        return nil;
    }

    if (![self message])
    {
        [self logAndAddReason:[NSString stringWithFormat:@"No embedded message in campaign %ld", (long)self.ID] withReasons:campaignReasons];
        return nil;
    }

    if (![self checkCampaignRulesForEvent:event atTime:time withReasons:campaignReasons]) {
        return nil;
    }

    return self.message;
}

#if TARGET_OS_IOS /** exclude tvOS **/
-(BOOL)supportsOrientation:(UIInterfaceOrientation)orientation {
    return YES;
}
#endif

-(BOOL)assetsReady:(NSSet *)assets withPersonalization:(NSDictionary *)personalization {
    return YES;
}

@end

File: public/SwrveSDKSwift/SwrveEmbeddedCampaign.swift
import Foundation
import UIKit

#if canImport(SwrveSDK)
import SwrveSDK
#endif

#if canImport(SwrveSDKCommon)
import SwrveSDKCommon
#endif

@objc public class SwrveEmbeddedCampaign: SwrveCampaign {

    /// Message attached to this embedded campaign.
    @objc public var message: SwrveEmbeddedMessage?

    /// Initializes the embedded campaign at the given time from the provided dictionary.
    ///
    /// - Parameters:
    ///   - time: The time of initialization.
    ///   - json: A dictionary containing the campaign data.
    /// - Returns: An initialized instance of `SwrveEmbeddedCampaign`.
    @objc public init(at time: Date, from json: [String: Any]) {
        super.init(at: time, from: json, campaignType: SWRVE_CAMPAIGN_EMBEDDED)

        if let embeddedMessageDict = json["embedded_message"] as? [String: Any] {
            self.message = SwrveEmbeddedMessage.fromDictionary(embeddedMessageDict, andCampaign: self)
        }

        self.name = self.message?.name ?? ""
    }

    /// Checks whether the campaign has a message for the given event with payload.
    ///
    /// - Parameters:
    ///   - event: The trigger event to check for.
    ///   - payload: A dictionary containing the payload for verifying conditions.
    /// - Returns: `true` if the campaign contains a message for the event, otherwise `false`.
    @objc public func hasMessage(forEvent event: String, withPayload payload: [String: Any]?) -> Bool {
        canTrigger(withEvent: event, andPayload: payload)
    }

    /// Retrieves the message for the given event, payload, and time.
    ///
    /// - Parameters:
    ///   - event: The trigger event.
    ///   - payload: The payload for verifying conditions.
    ///   - time: The current device time.
    ///   - campaignReasons: A dictionary containing the reasons the campaign did not return a message.
    /// - Returns: The embedded message for the event, or `nil` if no message exists.
    @objc public func message(
        forEvent event: String,
        withPayload payload: [String: Any]?,
        at time: Date,
        withReasons campaignReasons: NSMutableDictionary
    ) -> SwrveEmbeddedMessage? {

        if !hasMessage(forEvent: event, withPayload: payload) {
            let payloadDescription = SwrveUtilsSwift.formatPayloadForDisplay(payload)
            let reason = "There is no trigger in \(ID) that matches \(event) with conditions \(payloadDescription)"
            SwrveLogger.logDebug(reason)
            logAndAdd(reason: reason, withReasons: campaignReasons)
            return nil
        }

        guard let message = self.message else {
            let reason = "No embedded message in campaign \(self.ID)"
            logAndAdd(reason: reason, withReasons: campaignReasons)
            return nil
        }

        if !checkCampaignRules(forEvent: event, atTime: time, withReasons: campaignReasons) {
            return nil
        }

        return message
    }

    /// Checks whether the campaign supports a specific orientation.
    ///
    /// - Parameter orientation: The interface orientation.
    /// - Returns: `true` if the campaign supports the orientation, otherwise `false`.
    #if os(iOS)
    public override func supportsOrientation(_ orientation: UIInterfaceOrientation) -> Bool {
        true
    }
    #endif

    /// Checks if the assets for the campaign are ready.
    ///
    /// - Parameters:
    ///   - assets: The set of assets.
    ///   - personalization: The personalization dictionary.
    /// - Returns: `true` if the assets are ready, otherwise `false`.
    @objc public override func assetsReady(_ assets: Set<String>, withPersonalization personalization: [String: Any]?) -> Bool {
        true
    }
}
